compiler/noirc_frontend/src/hir_def/types.rs

```diff
@@ -1594,7 +1640,12 @@ impl Type {
                 // We may have already "bound" this type variable in this call to
                 // try_unify, so check those bindings as well.
                 match bindings.get(id) {
-                    Some((_, binding)) => binding.clone().try_unify(self, bindings),
+                    Some((_, kind, binding)) => {
+                        if !kind.unifies(&binding.kind()) {
+                            return Err(UnificationError);
+                        }
+                        binding.clone().try_unify(self, bindings)
+                    }
```

Comment:
I was thinking of if we could still perform this check if we removed kind from TypeBindings and I think we can since it should be the same kind as the type variable that was passed in, so we'd have to add a parameter.

Reply:
The `Kind` in `TypeBindings` can be more specific than the other `Kind`, e.g. without storing it in `TypeBindings`, we can lose the fact that we've specialized a `Kind::IntegerOrField` to a `Kind::Integer` during unification, and thus lose the ability to revert the type binding.